module/zfs/dsl_scan.c

```diff
@@ -762,6 +762,87 @@ dsl_scan_setup_sync(void *arg, dmu_tx_t *tx)
 	    *funcp, scn->scn_phys.scn_min_txg, scn->scn_phys.scn_max_txg);
 }
 
+static int
+dsl_rescrub_check(void *arg, dmu_tx_t *tx)
+{
+	dsl_scan_t *scn = dmu_tx_pool(tx)->dp_scan;
+ 	if (dsl_scan_is_running(scn))
+		return (SET_ERROR(EBUSY));
+	return (0);
+}
+
+static void
+dsl_rescrub_setup_sync(void *arg, dmu_tx_t *tx)
+{
+	uint64_t txg = tx->tx_txg;
+ 	dsl_scan_t *scn = dmu_tx_pool(tx)->dp_scan;
+ 	dsl_pool_t *dp = scn->scn_dp;
+	spa_t *spa = dp->dp_spa;
+ 	ASSERT(!dsl_scan_is_running(scn));
+	scn->scn_phys.scn_state = DSS_SCANNING;	
+
+ 	zfs_cmd_t zc = {"\0"};
+ 	uint64_t error_count = spa_get_errlog_size(spa);
+ 	if (error_count == 0){
+		scn->scn_phys.scn_state = DSS_FINISHED;
+		return;
+	}
+	zc.zc_nvlist_dst = (uintptr_t)kmem_zalloc(error_count * sizeof (zbookmark_phys_t), KM_SLEEP);
+	zc.zc_nvlist_dst_size = error_count;
+
+ 	size_t retrieved_error = (size_t)zc.zc_nvlist_dst_size;
+	int error = spa_get_errlog(spa, (void *)(uintptr_t)zc.zc_nvlist_dst,
+	    &retrieved_error);
+
+ 	if (error == 0){
+		zc.zc_nvlist_dst_size = retrieved_error;
+	}
+
+ 	zbookmark_phys_t *zb = NULL;
+	zb = ((zbookmark_phys_t *)(uintptr_t)zc.zc_nvlist_dst) +
+	    zc.zc_nvlist_dst_size;
+ 	error_count -= zc.zc_nvlist_dst_size;
+ 	// Erasing the previously on disk logged errors.
+	VERIFY(dmu_object_free(spa->spa_meta_objset, spa->spa_errlog_last, tx) == 0);
+ 	for (int i = 0; i < error_count; i++) {
+ 		dsl_dataset_t *ds;
+		int err = dsl_dataset_hold_obj(dp, zb[i].zb_objset, FTAG, &ds);
+ 		if (err) {
+			return;
+		}
+ 		objset_t *os = ds->ds_objset;
+		dmu_object_info_t doi;
+		dmu_object_info(os, zb[i].zb_object, &doi);
+
+ 		uint64_t indirect_block_size = doi.doi_metadata_block_size;
+ 		uint64_t blkptr_size = (uint64_t)sizeof (blkptr_t);
+		uint64_t data_block_size = doi.doi_data_block_size;
+		uint8_t indirect_block_shift = 0;
+		uint8_t blkptr_size_shift = 0;
+
+		while (indirect_block_size > 1) {
+			indirect_block_size = indirect_block_size >> 1;
+			indirect_block_shift++;
+		}
+		while (blkptr_size > 1) {
+			blkptr_size = blkptr_size >> 1;
+			blkptr_size_shift++;
+		}
+
+		uint64_t offset = data_block_size * (zb[i].zb_blkid << ((indirect_block_shift -
+		    blkptr_size_shift) * zb[i].zb_level));
+		uint64_t len = data_block_size * (1 << ((indirect_block_shift -
+		    blkptr_size_shift) * zb[i].zb_level));
+ 		dmu_prefetch(os, zb[i].zb_object, zb[i].zb_level, offset, len,
+		    ZIO_PRIORITY_NOW);
```

Comment:
It looks like you are prefetching each of the blocks in the error list.  We need to scrub them, which reads all DVA's from disk.

This is processing all the bookmarks on the error list in one txg.  If the error list is long, this could take a long time, negatively impacting performance.

Let's chat Monday about the design requirements and how we can accomplish them.